### PR TITLE
Add skip option for migration corrections

### PR DESCRIPTION
**Pull request by ylavoie**
- Add possibility to skip benign migration problems
- Optimize tooltips and button handling

### CODE REVIEW

File: lib/LedgerSMB/Scripts/setup.pm [hunk 1]
```diff
@@ -702,6 +702,8 @@ sub _applicable_upgrade_tests {
                   LedgerSMB::Upgrade_Tests->get_tests;
 }
 
+my @checks;
```
Comment (ehuelsmann) [inline lib/LedgerSMB/Scripts/setup.pm]: Our requests shouldn't depend on global state (I'm going through a lot of work to *eliminate* global state, such as App_State). Is there a reason you declare the checks as global?

File: lib/LedgerSMB/Upgrade_Tests.pm [hunk 2]
```diff
@@ -44,11 +44,8 @@ Returns the test object with the name.
 
 sub get_by_name {
     my ($self, $name) = @_;
-    my @tests = $self->_get_tests;
-    for my $test (@tests){
-       return $test if $test->name eq $name;
-    }
-    return;
+    my %tests = map { $_->name => $_ } $self->_get_tests;
+    return $tests{$name};
```
Comment (ehuelsmann) [inline lib/LedgerSMB/Upgrade_Tests.pm]: Why does this need to change? I mean, the old version, even if optimizing a non-bottleneck, wasn't exactly wrong...

Same functionality as the old version, no more values iterated through than required and nicely short:

```perl
use List::Util qw( first );

sub get_by_name {
   my ($self, $name) = @_;

   return first { $_->name eq $name } $self->_get_tests;
}
```

File: lib/LedgerSMB/Upgrade_Tests.pm [hunk 3]
```diff
@@ -122,6 +119,20 @@ C<text> is the textual value to be presented in the UI.
 
 has selectable_values => (is => 'ro', isa => 'HashRef', required => 0);
 
+sub _displayed_key {
```
Comment (ehuelsmann) [inline lib/LedgerSMB/Upgrade_Tests.pm]: The name suggests this returns displayed keys. However, from reading the code, that's not what it seems to do.

Do we really need this check? If so, maybe add a bit of documentation using `#` comments why this check is invaluable and what the exact check is achieving?

File: lib/LedgerSMB/Upgrade_Tests.pm [hunk 4]
```diff
@@ -122,6 +119,20 @@ C<text> is the textual value to be presented in the UI.
...
+    my ($self,@keys) = @_;
+    for my $key (@keys) {
+        die "'$key' not displayed in test '$self->{name}'"
+            if not grep( $key, $self->{display_cols});
+    }
+};
+
+after 'selectable_values' => sub {
+    my $self = shift;
+    $self->_displayed_key(keys %{$self->{selectable_values}})
```
Comment (ehuelsmann) [inline lib/LedgerSMB/Upgrade_Tests.pm]: Do we need to run this check on *every access* to `selectable_values`? or are you actually only trying to check it when being set?

File: lib/LedgerSMB/Upgrade_Tests.pm [hunk 5]
```diff
@@ -150,15 +161,16 @@ Repair query key to set the values if we can repair
 
 =cut
 
-has id_where => (is => 'ro', isa => 'Str', required => 0, default => 'id');
+has id_where => (is => 'ro', isa => 'Str', required => 0, default => '');
 
-=item id_column
+=item id_columns
```
Comment (ehuelsmann) [inline lib/LedgerSMB/Upgrade_Tests.pm]: You haven't changed all references to `id_column` to the new name (there's one left, as far as I can tell)

File: lib/LedgerSMB/Upgrade_Tests.pm [hunk 6]
```diff
@@ -168,6 +180,12 @@ Repair query columns to run once per result
 
 has columns => (is => 'ro', isa => 'ArrayRef[Str]', required => 0);
 
+after 'columns' => sub {
```
Comment (ehuelsmann) [inline lib/LedgerSMB/Upgrade_Tests.pm]: Same as with the selectable values:

1. do we really need the check?
2. and if we do: do we really need to do the check on every attribute access?

File: lib/LedgerSMB/Upgrade_Tests.pm [hunk 7]
```diff
@@ -214,13 +231,37 @@ Tooltip for each button
...
-        return {
-            'Save and Retry' => marktext('Save the fixes provided and attempt to continue migration'),
-            'Cancel' => marktext('Cancel the <b>migration</b>')
-    }},
-    required => 0);
+    isa => 'Maybe[HashRef[Str]]', required => 0,
+    default => undef,   # Force initializer call
+    initializer => sub {
+        my ( $self, $value, $writer_sub_ref, $attribute_meta ) = @_;
+        for ($value) {
```
Comment (ehuelsmann) [inline lib/LedgerSMB/Upgrade_Tests.pm]: `$value` is at best an array reference, but definitely not an array.

Please use a local value for iterating the array and explicitly iterate the array:

```perl
for my $btn (@$value) { };
```

File: lib/LedgerSMB/Upgrade_Tests.pm [hunk 8]
```diff
@@ -214,13 +231,37 @@ Tooltip for each button
...
+            die LedgerSMB::I18N::text(q(No button '[_1]' in test '[_2]'),$_,$self->{name})
```
Comment (ehuelsmann) [inline lib/LedgerSMB/Upgrade_Tests.pm]: Is this error detecting a programming error or providing the user with valuable feedback?

In case of the former, what are the chances of a true user ever seeing it? If the answer to that is "nil" then I don't think we should bother our translators with the need to translate it.

File: lib/LedgerSMB/Upgrade_Tests.pm [hunk 9]
```diff
@@ -214,13 +231,37 @@ Tooltip for each button
...
+        }
+        my %defaults = ('Save and Retry' => marktext('Save the fixes provided and attempt to continue migration'),
+                                'Cancel' => marktext('Cancel the <b>whole migration</b>'));
+        for (keys %defaults) {
+            $value->{$_} = $defaults{$_}
+                if not $value->{$_};
+        }
+        $writer_sub_ref->($value);
+    }
+);
```
Comment (ehuelsmann) [inline lib/LedgerSMB/Upgrade_Tests.pm]: The initializer feels over-engineered. How many more confirmations are we going to be seeing? Can we simply go for listing the confirmations that make sense for a specific test, deferring rendering of the buttons to the test-driver, supporting save(and retry), cancel and skip for now?

File: lib/LedgerSMB/Upgrade_Tests.pm [hunk 10]
```diff
@@ -214,13 +231,37 @@ Tooltip for each button
...
+
+=item skip
+
+Skip this test
+
+=cut
+
+enum SkipValues => [qw/ Disabled Off On /];
+has skip => (is =>'ro', isa => 'SkipValues',
```
Comment (ehuelsmann) [inline lib/LedgerSMB/Upgrade_Tests.pm]: This doesn't feel right: why do we have a special attribute to determine the status of exactly one of possibly infinite buttons?

File: lib/LedgerSMB/Scripts/setup.pm [hunk 11]
```diff
@@ -896,41 +893,45 @@ sub fix_tests{
...
+        $query = "UPDATE $table SET $setters WHERE $where "
+               . join(' AND ',map {"$_ = ? "} @{$check->id_columns});
     }
     my $sth = $dbh->prepare($query);
 
     for my $count (1 .. $request->{count}){
-        my $id = $request->{"id_$count"};
         my @values;
-        push @values, $id
+        push @values, split(/,/,@{$check->id_columns})
```
Comment (ehuelsmann) [inline lib/LedgerSMB/Scripts/setup.pm]: What's happening here? I thought the `id_columns` are an arrayref? Split is an operation that works on a string, not an array?

File: lib/LedgerSMB/Scripts/setup.pm [hunk 12]
```diff
@@ -896,41 +893,45 @@ sub fix_tests{
...
-        push @values, $request->{"id_$count"}
+        push @values, split(/,/,$request->{"id_$count"})
            if ! $check->{insert};
 
-        $sth->execute(@values) ||
+        my $rv = $sth->execute(@values) ||
             $request->error($sth->errstr);
+        #TODO: We should abort gracefully
+        die "$rv rows affected instead of only 1"
+            if $rv > 1;
```
Comment (ehuelsmann) [inline lib/LedgerSMB/Scripts/setup.pm]: Nicely defensive. I like it: it's not our own data, after all.

File: lib/LedgerSMB/Scripts/setup.pm [hunk 13]
```diff
@@ -1436,9 +1437,11 @@ Force work.  Forgets unmatching tests, applies a curing statement and move on.
 sub force{
```
Comment (ehuelsmann) [inline lib/LedgerSMB/Scripts/setup.pm]: I understand the Skip, Save and Cancel buttons, but what's the Force button for? I mean, we can't force a square peg in a round hole... If the schema forbids data to be stored, no amount of Force is going to help...?

File: lib/LedgerSMB/Upgrade_Tests.pm (hunk 2 above)
Comment (ylavoie) [reply lib/LedgerSMB/Upgrade_Tests.pm]: Ah! List::Util is core. So we won't pay for another cpan module.

File: lib/LedgerSMB/Upgrade_Tests.pm (hunk 3 above)
Comment (ylavoie) [reply lib/LedgerSMB/Upgrade_Tests.pm]: Renamed. Selectable values must be displayed otherwise test can't succeed.

File: lib/LedgerSMB/Upgrade_Tests.pm (hunk 4 above)
Comment (ylavoie) [reply lib/LedgerSMB/Upgrade_Tests.pm]: It should be in the builder, not on every access.

File: lib/LedgerSMB/Upgrade_Tests.pm (hunk 5 above)
Comment (ylavoie) [reply lib/LedgerSMB/Upgrade_Tests.pm]: Fixed

File: lib/LedgerSMB/Upgrade_Tests.pm (hunk 10 above)
Comment (ylavoie) [reply lib/LedgerSMB/Upgrade_Tests.pm]: Tests can be skippable. Simplified and renamed

File: lib/LedgerSMB/Scripts/setup.pm (hunk 13 above)
Comment (ylavoie) [reply lib/LedgerSMB/Scripts/setup.pm]: Sometimes, the peg problem is not the shape but the color. See the `no_businesses` test for an example. Either we force the user to go back in SL to define the missing business or we force remove the references to the missings.

Comment (ehuelsmann) [reply lib/LedgerSMB/Scripts/setup.pm]: Ok. I see now. So, `force` applies a query which looses the (non-conforming) data. That's clear now. Thanks for the `no_businesses` reference.

File: lib/LedgerSMB/Scripts/setup.pm [hunk 14]
```diff
@@ -779,72 +782,66 @@ sub _failed_check {
...
       push @$rows, $row;
-      my $count = scalar(@$rows);
-      $hiddens->{"id_$count"} = $row->{$check->id_column};
     }
     $hiddens->{count} = scalar(@$rows);
     $sth->finish();
 
     my $heading = { map { $_ => $_ } @{$check->display_cols} };
-    my %buttons = map { $_ => 1 } @{$check->buttons};
+    my %buttons_set = map { $_ => 1 } @{$check->buttons};
```
[the remark this answers is not in the document]
Comment (ylavoie) [reply lib/LedgerSMB/Scripts/setup.pm]: Fixed

File: lib/LedgerSMB/Scripts/setup.pm (hunk 11 above)
Comment (ylavoie) [reply lib/LedgerSMB/Scripts/setup.pm]: Fixed

File: lib/LedgerSMB/Scripts/setup.pm [hunk 15]
```diff
@@ -779,72 +782,65 @@ sub _failed_check {
     my $hiddens = {
        check => $check->name,
 verify_check => md5_hex($check->test_query),
-    database => $request->{database}
+    database => $request->{database},
```
Comment (ehuelsmann) [inline lib/LedgerSMB/Scripts/setup.pm]: If there's a style issue in the file (missing commas on last hash key/value pair), please change it in a separate commit. If there's no such style issue, why did this need to change in an otherwise unrelated change?

File: lib/LedgerSMB/Scripts/setup.pm [hunk 16]
```diff
@@ -779,72 +782,65 @@ sub _failed_check {
...
     };
+    my @skip_keys = grep /^skip_/, keys %$request;
+    $hiddens->{@skip_keys} = $request->{@skip_keys};
```
Comment (ehuelsmann) [inline lib/LedgerSMB/Scripts/setup.pm]: hash slices. Nice!

File: lib/LedgerSMB/Scripts/setup.pm [hunk 17]
```diff
@@ -779,72 +782,65 @@ sub _failed_check {
...
 
     my $rows = [];
     while (my $row = $sth->fetchrow_hashref('NAME_lc')) {
-      my $id = $row->{$check->{id_column}};
+      my $count = 1+scalar(@$rows);
+      #TODO: A '' will fail with invalid input syntax in fix_tests - YL
```
Comment (ehuelsmann) [inline lib/LedgerSMB/Scripts/setup.pm]: This comment is too terse for me. What is the problem you're trying to warn the reader for?

File: lib/LedgerSMB/Scripts/setup.pm [hunk 18]
```diff
@@ -779,72 +782,65 @@ sub _failed_check {
...
-    database => $request->{database}
+    database => $request->{database},
     };
+    my @skip_keys = grep /^skip_/, keys %$request;
+    $hiddens->{@skip_keys} = $request->{@skip_keys};
 
     my $rows = [];
     while (my $row = $sth->fetchrow_hashref('NAME_lc')) {
-      my $id = $row->{$check->{id_column}};
+      my $count = 1+scalar(@$rows);
```
Comment (ehuelsmann) [inline lib/LedgerSMB/Scripts/setup.pm]: `scalar(@$rows)` returns the exact number of rows in the `@$rows` array. Why do we need to add `1` to that?

File: lib/LedgerSMB/Scripts/setup.pm [hunk 19]
```diff
@@ -779,72 +782,65 @@ sub _failed_check {
...
+        my $name = $column . '_' . $count;
         $row->{$column} =
            ( defined $selectable_value && @$selectable_value )
            ? { select => {
-                   name => $column . "_$id",
-                   id => $id,
+                   name => $name,
+                   id => $count,
                    options => $selectable_value,
                    default_blank => ( 1 != @$selectable_value )
```
Comment (ehuelsmann) [inline lib/LedgerSMB/Scripts/setup.pm]: Shouldn't the value of the dropdown be set, just as with the input field below, based on the value that can be found in `$row->{$column}`?

File: lib/LedgerSMB/Scripts/setup.pm [hunk 20]
```diff
@@ -779,72 +782,65 @@ sub _failed_check {
...
       for my $column (@{$check->columns // []}) {
         my $selectable_value = $selectable_values{$column};
+        my $name = $column . '_' . $count;
         $row->{$column} =
            ( defined $selectable_value && @$selectable_value )
            ? { select => {
-                   name => $column . "_$id",
-                   id => $id,
+                   name => $name,
+                   id => $count,
```
Comment (ehuelsmann) [inline lib/LedgerSMB/Scripts/setup.pm]: The previous `$id` value was different for each row. The new `$count` value doesn't seem to be. Am I missing something?

File: lib/LedgerSMB/Scripts/setup.pm [hunk 21]
```diff
@@ -779,72 +782,65 @@ sub _failed_check {
...
 
     my $rows = [];
     while (my $row = $sth->fetchrow_hashref('NAME_lc')) {
-      my $id = $row->{$check->{id_column}};
+      my $count = 1+scalar(@$rows);
+      #TODO: A '' will fail with invalid input syntax in fix_tests - YL
+      my $id = join(',',map { $row->{$_} // '' } @{$check->id_columns});
       for my $column (@{$check->columns // []}) {
         my $selectable_value = $selectable_values{$column};
+        my $name = $column . '_' . $count;
```
Comment (ehuelsmann) [inline lib/LedgerSMB/Scripts/setup.pm]: This makes `$name` to be the same value for all cells in a single column, right?

File: lib/LedgerSMB/Scripts/setup.pm [hunk 22]
```diff
@@ -779,72 +782,65 @@ sub _failed_check {
...
     };
+    my @skip_keys = grep /^skip_/, keys %$request;
+    $hiddens->{@skip_keys} = $request->{@skip_keys};
 
     my $rows = [];
     while (my $row = $sth->fetchrow_hashref('NAME_lc')) {
-      my $id = $row->{$check->{id_column}};
+      my $count = 1+scalar(@$rows);
+      #TODO: A '' will fail with invalid input syntax in fix_tests - YL
+      my $id = join(',',map { $row->{$_} // '' } @{$check->id_columns});
```
Comment (ehuelsmann) [inline lib/LedgerSMB/Scripts/setup.pm]: What happens when the value in the primary key is a character field containing a comma? (Or do we only have numeric p-keys so far?)

File: lib/LedgerSMB/Scripts/setup.pm [hunk 23]
```diff
@@ -779,72 +782,66 @@ sub _failed_check {
...
+                           msg => $check->{tooltips}->{$_->{label}}
+                                ? $request->{_locale}->maketext($check->{tooltips}->{$_->{label}})
+                                : undef,
+                           position => 'above'
+                         },
+                 text => $request->{_locale}->maketext($_->{label}),
+                class => 'submit'
+            }
+        }
+    }
```
[the remark this answers is not in the document]
Comment (ylavoie) [reply lib/LedgerSMB/Scripts/setup.pm]: Thanks. After the introduction of the last button, there really wasn't any other choice.

File: lib/LedgerSMB/Scripts/setup.pm [hunk 24]
```diff
@@ -896,41 +892,45 @@ sub fix_tests{
...
+    my $id_columns = join('|',@{$check->id_columns});
     my $id_displayed = $check->{insert}
-                and grep( /^$check->{id_column}$/, @{$check->display_cols} );
+                and grep( /^($id_columns)$/, @{$check->display_cols} );
 
     my $query;
     if ($check->{insert}) {
         my @_edits = @edits;
-        unshift @_edits, $check->{id_column} if $id_displayed;
+        unshift @_edits, @{$check->id_columns} if $id_displayed;
```
Comment (ehuelsmann) [inline lib/LedgerSMB/Scripts/setup.pm]: I'm unfamiliar with the UI, but why doesn't the UI provide us with these values? Isn't the key to the situation that the row is *new*, so the p-key isn't available in advance?

File: lib/LedgerSMB/Scripts/setup.pm [hunk 25]
```diff
@@ -1436,9 +1436,11 @@ Force work.  Forgets unmatching tests, applies a curing statement and move on.
 sub force{
     my ($request) = @_;
     my $database = _init_db($request);
+    my $dbinfo = $database->get_info();
 
-    my %test = map { $_->name => $_ } LedgerSMB::Upgrade_Tests->get_tests();
-    my $force_queries = $test{$request->{check}}->{force_queries};
+    my %checks = map { $_->name => $_ } @{$info_applicable_for_upgrade{$dbinfo}};
```
Comment (ehuelsmann) [inline lib/LedgerSMB/Scripts/setup.pm]: I think this is wrong. You probably meant to call `applicable_for_upgrade()` instead of evaluating the hash `applicable_for_upgrade`.

File: lib/LedgerSMB/Scripts/setup.pm [hunk 26]
```diff
@@ -1436,9 +1436,11 @@ Force work.  Forgets unmatching tests, applies a curing statement and move on.
...
+    my $check = $checks{$request->{check}};
```
Comment (ehuelsmann) [inline lib/LedgerSMB/Scripts/setup.pm]: This statement and the one above can be done with `first()` from `List::Util` again.

File: lib/LedgerSMB/Upgrade_Tests.pm [hunk 27]
```diff
@@ -150,15 +163,16 @@ Repair query key to set the values if we can repair
 
 =cut
 
-has id_where => (is => 'ro', isa => 'Str', required => 0, default => 'id');
+has id_where => (is => 'ro', isa => 'Str', required => 0, default => '');
```
Comment (ehuelsmann) [inline lib/LedgerSMB/Upgrade_Tests.pm]: Now that we support complex primary keys, what value does this attribute still have?

File: lib/LedgerSMB/Upgrade_Tests.pm [hunk 28]
```diff
@@ -200,27 +214,69 @@ Enabled buttons
...
+
+=item BUILD
+
+Called after object building. Ideal to validate object fields for coherence
+
+=cut
+
+sub BUILD {
+    my $self = shift;
+    $self->_validate_displayed_key(keys %{$self->{selectable_values}})
```
Comment (ehuelsmann) [inline lib/LedgerSMB/Upgrade_Tests.pm]: This check is done at run time (when the user is upgrading). However, it's validating information that's statically available in this file. For example the `no_business_for_customer` check defines all the components required to run this check. So, even without having a database to migrate, we can just check the column-consistency when running our `prove` tests.
And once we have established our `prove` outcome to be correct, why would we still want to run this check during a migration?

File: lib/LedgerSMB/Scripts/setup.pm (hunk 18 above)
Comment (ylavoie) [reply lib/LedgerSMB/Scripts/setup.pm]: Because the rows are empty initialy and id starts at 1.

File: lib/LedgerSMB/Scripts/setup.pm (hunk 22 above)
Comment (ylavoie) [reply lib/LedgerSMB/Scripts/setup.pm]: ID of columns are SQL identifiers

File: lib/LedgerSMB/Scripts/setup.pm (hunk 21 above)
Comment (ylavoie) [reply lib/LedgerSMB/Scripts/setup.pm]: No, `$count` changes everytime we push a line in `$rows`

File: lib/LedgerSMB/Scripts/setup.pm (hunk 20 above)
Comment (ylavoie) [reply lib/LedgerSMB/Scripts/setup.pm]: See previous comment

File: lib/LedgerSMB/Scripts/setup.pm (hunk 19 above)
Comment (ylavoie) [reply lib/LedgerSMB/Scripts/setup.pm]: We never did because if the data was presented to the user to fix, then the original value was wrong.

File: lib/LedgerSMB/Scripts/setup.pm (hunk 25 above)
Comment (ylavoie) [reply lib/LedgerSMB/Scripts/setup.pm]: Fixed

File: lib/LedgerSMB/Upgrade_Tests.pm (hunk 28 above)
Comment (ylavoie) [reply lib/LedgerSMB/Upgrade_Tests.pm]: Moved to `xt/50-upgrade-tests.t`. Same for correlation between tooltips and buttons.

File: lib/LedgerSMB/Scripts/setup.pm (hunk 22 above)
Comment (ylavoie) [reply lib/LedgerSMB/Scripts/setup.pm]: id_columns contains the columns id, not data. So, unless we want to protect ourselves in case of someone starts to use quoted ids, which we don't elsewhere, joining the id array with commas is totally acceptable.

Comment (ehuelsmann) [reply lib/LedgerSMB/Scripts/setup.pm]: Sure. But `map { $row->{$_} // '' }` maps the column names into the data for the current record. Which means you can get any content, including commas.

File: lib/LedgerSMB/Upgrade_Tests.pm (hunk 27 above)
Comment (ylavoie) [reply lib/LedgerSMB/Upgrade_Tests.pm]: None. Either id_where or id_columns can be used. Both cannot be empty though.

Comment (ehuelsmann) [reply lib/LedgerSMB/Upgrade_Tests.pm]: I don't understand why we need `id_where` in case we already have `id_columns`

File: lib/LedgerSMB/Scripts/setup.pm [hunk 29]
```diff
@@ -896,41 +895,45 @@ sub fix_tests{
...
+          push @values, $request->{"${edit}_$count"};
         }
-        push @values, $request->{"id_$count"}
+        push @values, map { MIME::Base64::decode($_)} split(/,/,$request->{"id_$count"})
            if ! $check->{insert};
 
-        $sth->execute(@values) ||
+        my $rv = $sth->execute(@values) ||
             $request->error($sth->errstr);
+        #TODO: We should abort gracefully
```
Comment (ehuelsmann) [inline lib/LedgerSMB/Scripts/setup.pm]: No need to `die()` here: simply return a PSGI triplet with a body indicating the problem.
